### PR TITLE
Add findByCurrentStatus method with sort parameter to order by contributionDeadline

### PR DESCRIPTION
**Pull request by jbern0rd**
Draft PR for comments
To be complete, this feature would merit a basic integration test to check the mongodb query is correct and returns tasks in the expected order.

### CODE REVIEW

File: src/main/java/com/iexec/core/task/TaskRepository.java [hunk 1]
```diff
@@ -40,6 +41,9 @@
     @Query("{ 'currentStatus': {$in: ?0} }")
     List<Task> findByCurrentStatus(List<TaskStatus> statuses);
 
+    @Query("{ 'currentStatus'; {$in: ?0} }")
```
Comment (zguesmi) [inline src/main/java/com/iexec/core/task/TaskRepository.java:44]:
```suggestion
    @Query("{ 'currentStatus': {$in: ?0} }")
```

File: src/test/resources/logback.xml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+<configuration>
+
+    <!-- Default -->
+    <property name="LOG_LEVEL" value="${IEXEC_LOG_LEVEL:-INFO}"/>
+    <include resource="org/springframework/boot/logging/logback/base.xml"/>
+    <logger name="org.springframework.web" level="${LOG_LEVEL}"/>
+
+    <!-- iexec -->
+    <logger name="com.iexec" leve="DEBUG"/>
```
Comment (mcornaton) [inline src/test/resources/logback.xml:26]: Shouldn't this be `level` instead of `leve`?

Comment (jbern0rd) [reply src/test/resources/logback.xml]: Fixed, thanks

File: src/main/java/com/iexec/core/chain/DealWatcherService.java [hunk 3]
```diff
@@ -59,11 +61,12 @@ public DealWatcherService(IexecHubService iexecHubService,
     }
 
     /**
-     * This should be non-blocking to librate
+     * This should be non-blocking to liberate
      * the main thread, since deals can have
      * a large number of tasks (BoT).
      */
     @Async
+    @EventListener(ApplicationReadyEvent.class)
```
Comment (jeremyjams) [inline src/main/java/com/iexec/core/chain/DealWatcherService.java:69]: Nice!

File: src/main/java/com/iexec/core/task/TaskService.java [hunk 4]
```diff
@@ -109,7 +110,8 @@ public TaskService(TaskRepository taskRepository) {
     }
 
     public List<Task> getInitializedOrRunningTasks() {
-        return taskRepository.findByCurrentStatus(Arrays.asList(INITIALIZED, RUNNING));
+        return taskRepository.findByCurrentStatus(Arrays.asList(INITIALIZED, RUNNING),
+                Sort.by(Sort.Direction.ASC, "contributionDeadline"));
```
Comment (jeremyjams) [inline src/main/java/com/iexec/core/task/TaskService.java:114]: How about something like:
`Sort.by(Sort.Direction.ASC, Task.CONTRIBUTION_DEADLINE_FIELD_NAME)`
```
@Field(CONTRIBUTION_DEADLINE_FIELD_NAME)
private Date contributionDeadline;
```
 
 To protect against future updates?

**Review by zguesmi — APPROVED** [review]
Well done :clap:

File: src/main/java/com/iexec/core/task/TaskService.java (hunk 4 above)
Comment (jbern0rd) [reply src/main/java/com/iexec/core/task/TaskService.java]: Fixed, thanks for the suggestion !

**Review by jeremyjams — APPROVED** [review]
Great, thank you for your tests!

**Review by mcornaton — APPROVED** [review]
Thanks!

File: src/test/java/com/iexec/core/task/TaskRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            Task task = getStubTask(maxExecutionTime);
+            task.setChainDealId(generateChainId());
+            task.setChainTaskId(generateChainId());
+            task.setCurrentStatus(generator.nextInt(50) % 2 == 0 ? RUNNING : INITIALIZED);
+            int amountToAdd = generator.nextInt(10);
+            task.setContributionDeadline(Date.from(Instant.now().plus(amountToAdd, ChronoUnit.MINUTES)));
+            tasks.add(task);
+        }
+        taskRepository.saveAll(tasks);
+        tasks.sort(Comparator.comparing(Task::getCurrentStatus, Comparator.reverseOrder())
```
Comment (jeremyjams) [inline src/test/java/com/iexec/core/task/TaskRepositoryTest.java:101]: I think it is interesting/clearer to add at least 1 test where expected task list is explicitly defined.

File: src/test/java/com/iexec/core/task/TaskRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                .hasCauseExactlyInstanceOf(com.mongodb.MongoBulkWriteException.class)
+                .hasMessageContaining("duplicate key error collection: iexec.task index: chainTaskId dup key");
+    }
+
+    @Test
+    void shouldFindTasksOrderedByCurrentStatusAndContributionDeadline() {
+        List<Task> tasks = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            Task task = getStubTask(maxExecutionTime);
+            task.setChainDealId(generateChainId());
```
Comment (jeremyjams) [inline src/test/java/com/iexec/core/task/TaskRepositoryTest.java:93]:
```suggestion
            task.setChainDealId(generateHexId());
```

File: src/test/java/com/iexec/core/task/TaskRepositoryTest.java (hunk 5 above)
Comment (jbern0rd) [reply src/test/java/com/iexec/core/task/TaskRepositoryTest.java:101]: Would it be alright to have a simple test with fixed data and a fuzzy one with randomization to try different combinations ?

Comment (jeremyjams) [reply src/test/java/com/iexec/core/task/TaskRepositoryTest.java:101]: That would be perfect, having both benefits!

File: src/test/java/com/iexec/core/task/TaskRepositoryTest.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.*;
+
+import static com.iexec.core.task.TaskStatus.INITIALIZED;
+import static com.iexec.core.task.TaskStatus.RUNNING;
+import static com.iexec.core.task.TaskTestsUtils.getStubTask;
+
+@DataMongoTest
```
Comment (mcornaton) [inline src/test/java/com/iexec/core/task/TaskRepositoryTest.java:25]: Are we sure we want Spring tests in our unit tests?

Comment (jbern0rd) [reply src/test/java/com/iexec/core/task/TaskRepositoryTest.java:25]: Those are specific tests which only create beans for entities and repositories. The beans annotated with @Components and @Services are not created.
The new tests specifically target and validate the data model which should be validated as early as possible.
There is a valid discussion on the topic and we decided to start this way and maybe to refactor later.

**Pull request merged**